### PR TITLE
Client callbacks

### PR DESCRIPTION
**Pull request by minglis**
I'm not in today so this could wait till Monday for demos, raising pull request for code review only, and to park this work.

Commented the files a bit in this pull request to help reviews.

This pull request wires in the callbacks from Firetext and AWS SES. In both cases when a message has been processed by those platforms they call back to a URL on Notify to update the status:

Firetext:
- Update to SMS task to send notification id as a "reference". This is passed back in the callback to allow status updates
- Validation codes use the 'send-sms-code' as the reference to allow us to know whether to attempt a notification update
- Firetext client has an object to map firetext status codes to notifcation states.
- New endpoint in the notification rest controller - POST /notifications/sms/firetext
- Fire text response is a from encoded string (example curl)
`curl -X POST http://localhost:6011/notifications/sms/firetext -d'mobile=441234123123&status=0&reference=145ad964-d0e3-476b-ae08-7ed1bbb43612'`
- Endpoint does a large amount of logging - this is because we can't wire this in locally and we may need some help debugging on preview.

SES:
- These is no reference on SES so there is an update to the Notification object to store the 'reference' which is the messageId from SES, generated as part of the API call to AWS.
- This is then used to update the notification as part of the SES callback
- SES callback is a big blob of JSON. There is again a lot of debug / logging in the endpoint to help deployment to AWS.
- Same principles as in fire text, new endpoint (POST /notifications/email/ses), example curl

` curl -H "Content-type: text/plain; charset=UTF-8" -X POST http://localhost:6011/notifications/email/ses -d'{ "MessageId": "35efe472-ba36-5808-89bb-ab2925158b13", "UnsubscribeURL": "https://sns.eu-west-1.amazonaws.com/?Action=Unsubscribe&SubscriptionArn=arn:aws:sns:eu-west-1:123456789012:preview-emails:12345678-1234-1234-1234-123456789012", "SignatureVersion": "1", "TopicArn": "arn:aws:sns:eu-west-1:123456789012:testing", "Timestamp": "2016-03-10T16:12:19.876Z", "Type": "Notification", "Signature": "sig", "Message": { "notificationType": "Delivery", "mail": { "timestamp": "2016-03-10T16:12:19.016Z", "source": "invites@testing-notify.com", "sourceArn": "arn:aws:ses:eu-west-1:123456789012:identity/testing-notify", "sendingAccountId": "123456789012", "messageId": "01020153614cd6c8-2ec4bd32-7ddc-4344-811e-65d05519251f-000000", "destination": ["+447827992607"] }, "delivery": { "timestamp": "2016-03-10T16:12:19.751Z", "processingTimeMillis": 735, "recipients": ["testing@digital.cabinet-office.gov.uk"], "smtpResponse": "250 2.0.0 OK 1457626339 u62si5491824wme.91 - gsmtp", "reportingMTA": "a6-15.smtp-out.eu-west-1.amazonses.com" } }, "SigningCertURL": "https://sns.eu-west-1.amazonaws.com/SimpleNotificationService-bb750dd426d95ee9390147a5624348ee.pem" }'
`

- For emails we use the source field to identify invites/validation codes. If the email was sent from these accounts we don't try and update the notification.

Using these curls we can test locally - update messageId/reference to match notifications your local schema.

### CODE REVIEW

File: app/__init__.py [hunk 1]
```diff
@@ -71,7 +71,12 @@ def create_app():
 def init_app(app):
     @app.before_request
     def required_authentication():
-        if request.path != url_for('status.show_status'):
+        no_auth_req = [
```
Comment (minglis) [inline app/__init__.py]: Currently the callbacks are not authed. I think AWS supports this - I need to talk to firetext and other partners to see if they do to.

Note the risk is small (need correct JSON/ID's to make an update) but not ideal.

File: app/celery/tasks.py [hunk 2]
```diff
@@ -203,10 +204,14 @@ def send_sms(service_id, notification_id, encrypted_notification, created_at):
                 )
 
                 client.send_sms(
-                    notification['to'],
-                    template.replaced
+                    to=notification['to'],
+                    content=template.replaced,
+                    reference=str(notification_id)
```
Comment (minglis) [inline app/celery/tasks.py]: Using the notification id as the reference,

File: app/celery/tasks.py [hunk 3]
```diff
@@ -269,12 +274,13 @@ def send_email(service_id, notification_id, subject, from_address, encrypted_not
...
                 )
 
-                client.send_email(
+                reference = client.send_email(
                     from_address,
                     notification['to'],
                     subject,
                     template.replaced
                 )
+                update_notification_reference_by_id(notification_id, reference)
```
Comment (minglis) [inline app/celery/tasks.py]: Update the notification once the email has been sent to AWS. Wanted to minimse the DB writes here, but I think the update is OK here as we need to save the notification prior to the AWS call, and the ID is only available after,

File: app/celery/tasks.py [hunk 4]
```diff
@@ -291,7 +297,9 @@ def send_email(service_id, notification_id, subject, from_address, encrypted_not
 def send_sms_code(encrypted_verification):
     verification_message = encryption.decrypt(encrypted_verification)
     try:
-        firetext_client.send_sms(verification_message['to'], verification_message['secret_code'])
+        firetext_client.send_sms(
+            verification_message['to'], verification_message['secret_code'], 'send-sms-code'
```
Comment (minglis) [inline app/celery/tasks.py]: String "send-sms-code" to identify the firetext callback as different to a notification one.

File: app/clients/email/aws_ses.py [hunk 5]
```diff
@@ -3,6 +3,21 @@
 from monotonic import monotonic
 from app.clients.email import (EmailClientException, EmailClient)
 
+ses_response_status = {
```
Comment (minglis) [inline app/clients/email/aws_ses.py]: SES status. These are all now valid status for a notification DB object too.

File: app/clients/sms/firetext.py [hunk 6]
```diff
@@ -9,9 +9,32 @@
 
 logger = logging.getLogger(__name__)
 
+firetext_response_status = {
```
Comment (minglis) [inline app/clients/sms/firetext.py]: Firetext status. Mapped to notify status. Note "undelivered/pending" is an odd state and not sure what we should store that as, if we'd get another callback to update to delivered. Again a question for Firetext. (sent).

File: app/clients/sms/firetext.py [hunk 7]
```diff
@@ -28,22 +51,26 @@ def init_app(self, config, *args, **kwargs):
...
+            "reference": reference
         }
 
+        start_time = monotonic()
         try:
-            start_time = monotonic()
             response = request(
                 "POST",
-                "https://www.firetext.co.uk/api/sendsms",
+                "https://www.firetext.co.uk/api/sendsms/json",
```
Comment (minglis) [inline app/clients/sms/firetext.py]: Get JSON responses.

File: app/clients/sms/firetext.py [hunk 8]
```diff
@@ -28,22 +51,26 @@ def init_app(self, config, *args, **kwargs):
...
                 data=data
             )
+            firetext_response = response.json()
+            if firetext_response['code'] != 0:
+                raise FiretextClientException(firetext_response)
```
Comment (minglis) [inline app/clients/sms/firetext.py]: Firetext exception to capture response. Often a status code is not used, so inspect the status code and raise if non-zero.

File: app/dao/notifications_dao.py [hunk 9]
```diff
@@ -64,10 +64,44 @@ def update_job_sent_count(notification):
 
 
 def dao_update_notification(notification):
+    notification.updated_at = datetime.utcnow()
     db.session.add(notification)
     db.session.commit()
 
 
+def update_notification_status_by_id(notification_id, status):
```
Comment (minglis) [inline app/dao/notifications_dao.py]: Repetitive update methods - to avoid serialising objects for speed/memory reasons.

File: app/models.py [hunk 10]
```diff
@@ -231,7 +231,7 @@ def check_code(self, cde):
         return check_hash(cde, self._code)
 
 
-NOTIFICATION_STATUS_TYPES = ['sent', 'failed']
+NOTIFICATION_STATUS_TYPES = ['sent', 'delivered', 'failed', 'complaint', 'bounce']
```
Comment (minglis) [inline app/models.py]: SES states.

File: app/notifications/rest.py [hunk 11]
```diff
@@ -32,6 +35,150 @@
 register_errors(notifications)
 
 
+@notifications.route('/notifications/email/ses', methods=['POST'])
```
Comment (minglis) [inline app/notifications/rest.py]: Process SES callback. Lots of if statements as we inspect the JSON etc. Probably a bit wordy but as unable to wire up locally thought we may need some info on the preview/staging env.

File: app/notifications/rest.py [hunk 12]
```diff
@@ -32,6 +35,150 @@
...
+        current_app.config['NOTIFY_EMAIL_DOMAIN']
+    )
+    if current_app.config['VERIFY_CODE_FROM_EMAIL_ADDRESS'] == source:
+        return True
+    if invite_email == source:
+        return True
+    return False
+
+
+@notifications.route('/notifications/sms/firetext', methods=['POST'])
```
Comment (minglis) [inline app/notifications/rest.py]: Firetext response - as SES one in terms of debugging.

File: test_ses_responses/ses_response.json [hunk 13]
```diff
@@ -0,0 +1,32 @@
+{
```
Comment (minglis) [inline test_ses_responses/ses_response.json]: Big JSON response from SES. Built as a file so easier to read.

Comment (minglis) [inline test_ses_responses/ses_response.json]: All real IDs stripped.

File: app/celery/tasks.py [hunk 14]
```diff
@@ -269,12 +274,13 @@ def send_email(service_id, notification_id, subject, from_address, encrypted_not
...
-                client.send_email(
+                reference = client.send_email(
                     from_address,
                     notification['to'],
                     subject,
                     template.replaced
                 )
+                update_notification_reference_by_id(notification_id, reference)
             except AwsSesClientException as e:
                 current_app.logger.debug(e)
```
Comment (mannickutd) [inline app/celery/tasks.py]: For the firetext client exception debug level is used and here debug?

File: app/dao/notifications_dao.py [hunk 15]
```diff
@@ -64,10 +64,44 @@ def update_job_sent_count(notification):
...
+    db.session.commit()
+    return count
+
+
+def update_notification_reference_by_id(id, reference):
+    count = db.session.query(Notification).filter_by(
+        id=id
+    ).update({
+        Notification.reference: reference,
+        Notification.updated_at: datetime.utcnow()
```
Comment (mannickutd) [inline app/dao/notifications_dao.py]: The db model for notification uses an 'on update' clause to update the updated_at field.

File: app/notifications/rest.py [hunk 16]
```diff
@@ -32,6 +35,150 @@
...
+
+
+def is_not_a_notification(source):
+    invite_email = "{}@{}".format(
+        current_app.config['INVITATION_EMAIL_FROM'],
+        current_app.config['NOTIFY_EMAIL_DOMAIN']
+    )
+    if current_app.config['VERIFY_CODE_FROM_EMAIL_ADDRESS'] == source:
+        return True
+    if invite_email == source:
```
Comment (mannickutd) [inline app/notifications/rest.py]: Can we not combine the 2 if statements.

File: app/notifications/rest.py [hunk 17]
```diff
@@ -32,6 +35,150 @@
...
+
+@notifications.route('/notifications/sms/firetext', methods=['POST'])
+def process_firetext_response():
+    if 'status' not in request.form:
+        current_app.logger.info(
+            "Firetext callback failed: status missing"
+        )
+        return jsonify(result="error", message="Firetext callback failed: status missing"), 400
+
+    if len(request.form.get('reference', '')) <= 0:
```
Comment (mannickutd) [inline app/notifications/rest.py]: This is very java ish.
if not request.form.get('reference', None)??

File: tests/app/dao/test_notification_dao.py [hunk 18]
```diff
@@ -568,7 +600,7 @@ def test_should_not_delete_sent_notifications_before_one_day(notify_db, notify_d
 
 def test_should_not_delete_failed_notifications_before_seven_days(notify_db, notify_db_session):
     expired = datetime.utcnow() - timedelta(hours=24 * 7)
-    valid = datetime.utcnow() - timedelta(hours=(24 * 6) + 23,  minutes=59, seconds=59)
+    valid = datetime.utcnow() - timedelta(hours=(24 * 6) + 23, minutes=59, seconds=59)
```
Comment (mannickutd) [inline tests/app/dao/test_notification_dao.py]: Possibly one for the future reference if we do more and more datetime handling.
http://dateutil.readthedocs.org/en/latest/index.html

File: app/dao/notifications_dao.py (hunk 15 above)
Comment (minglis) [inline app/dao/notifications_dao.py]: Noticed that it wasn't updating in the DB? So started doing it by hand.

**Comment by quis** [comment]
Seen demo. :+1: to merge when comments addressed and merge conflicts fixed.

File: app/notifications/rest.py (hunk 16 above)
Comment (minglis) [inline app/notifications/rest.py]: It just looked really evil like that, thought might be cleaner? Happy to change if not.

File: app/notifications/rest.py (hunk 17 above)
Comment (minglis) [inline app/notifications/rest.py]: I'm checking here for presence/missing and present/empty - as it looks like both are possible.

File: app/celery/tasks.py (hunk 14 above)
Comment (minglis) [inline app/celery/tasks.py]: Fixed - both at exception
